### ISSUE TITLE
InvalidChannelState(Error)

### ISSUE THREAD

**Issue by alekspickle** [amqp-rs/lapin#392]
I am experiencing something in #330 but it's weird: I'm using `testcontainers` to setup rabbit container (super convenient for integration tests) but for some reason my setup fails:

<details>
<summary>MRE</summary>

```
#Cargo.toml
[dependencies]
env_logger = "0.10.0"
lapin = "2.3.1"
log = "0.4.20"
testcontainers = "0.15.0"
tokio = { version = "1", features = ["full"] }

```

```
// main.rs

use lapin::{options::ExchangeDeclareOptions, Connection, ExchangeKind};
pub use testcontainers::{clients::Cli, core::WaitFor, Container, GenericImage, RunnableImage};

const PORT: u16 = 5672;

#[tokio::main]
async fn main() {
    env_logger::init();

    let docker = Cli::default();
    let container = docker.run(
        GenericImage::new("rabbitmq", "3.12-management")
            .with_wait_for(WaitFor::message_on_stdout("Server startup complete;"))
            .with_exposed_port(PORT),
    );
    println!("Connect with rabbit");
    let port = container.get_host_port_ipv4(PORT);
    let connection = Connection::connect(&format!("amqp://127.0.0.1:{port}/%2f"), <_>::default())
        .await
        .unwrap();

    let channel = connection.create_channel().await.expect("channel1");

    println!("Declare exchange");
    // expect this to fail as no such exchange exists
    // we also expect the channel to be dead now
    channel
        .exchange_declare(
            "foo-exchange",
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                passive: true,
                ..<_>::default()
            },
            <_>::default(),
        )
        .await
        .expect("declare exchange");

    let _new_channel = connection.create_channel().await.expect("channel2");

    // non-passively declare exchange etc
}

```

</details>

the dfference between lapin 1.6.9 and 2.3.1

```
// 1.6.9 
Connect with rabbit
Declare exchange
[2023-10-06T11:53:50Z ERROR lapin::channel] Channel closed on channel 1 by 40:10 => AMQPError { kind: Soft(NOTFOUND), message: ShortString("NOT_FOUND - no exchange 'foo-exchange' in vhost '/'") } => NOT_FOUND - no exchange 'foo-exchange' in vhost '/'
[2023-10-06T11:53:50Z ERROR lapin::channels] Connection error: invalid channel state: Error
thread 'main' panicked at 'declare exchange: ProtocolError(AMQPError { kind: Soft(NOTFOUND), message: ShortString("NOT_FOUND - no exchange 'foo-exchange' in vhost '/'") })', src/main.rs:38:10

// 2.3.1
Connect with rabbit
Declare exchange
thread 'main' panicked at 'declare exchange: ProtocolError(AMQPError { kind: Soft(NOTFOUND), message: ShortString("NOT_FOUND - no exchange 'foo-exchange' in vhost '/'") })', src/main.rs:38:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


```

Is it the same issue and in the newer version there is no root cause for exchange declare fail or is it some other issue?

**Comment by Keruspe** [comment]
I don't see an InvalidChannelState with up to date lapin ?

You are setting passive to true, which means declares does nothing if it already exists, and an error is raised if it does not exist. This is what passive means, in AMQP

**Comment by alekspickle** [comment]
Thx for the fast response!
Regardless of exchange, can you explain this channel state change when it is passed from the function?

<details>
<summary>code with prints</summary>

```

use lapin::{options::ExchangeDeclareOptions, Channel, Connection, ExchangeKind};
pub use testcontainers::{clients::Cli, core::WaitFor, Container, GenericImage, RunnableImage};

const PORT: u16 = 5672;

#[tokio::main]
async fn main() {
    // expect this to fail as no such exchange exists
    // we also expect the channel to be dead now
    let (con, channel) = setup().await;
    println!("Channel state after setup:{:?}", channel.status().state());
    channel
        .exchange_declare(
            "foo-exchange",
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                passive: true,
                ..<_>::default()
            },
            <_>::default(),
        )
        .await
        .expect("declare exchange");

}

async fn setup() -> (Connection, Channel) {
    env_logger::init();

    let docker = Cli::default();
    let container = docker.run(
        GenericImage::new("rabbitmq", "3.12-management")
            .with_wait_for(WaitFor::message_on_stdout("Server startup complete;"))
            .with_exposed_port(PORT),
    );
    println!("Connect with rabbit");
    let port = container.get_host_port_ipv4(PORT);
    let con = Connection::connect(&format!("amqp://127.0.0.1:{port}/%2f"), <_>::default())
        .await
        .unwrap();

    let channel = con.create_channel().await.expect("channel1");
    println!("Channel state during setup:{:?}", channel.status().state());

    (con, channel)
}


```

</details>

with 2.3.1 I get

```

Connect with rabbit
Channel state during setup:Connected
Channel state after setup:Error
thread 'main' panicked at 'declare exchange: InvalidChannelState(Error)', src/main.rs:24:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


```

**Comment by alekspickle** [comment]
@Keruspe thx for the fast response!
Regardless of exchange, can you explain this channel state change when it is passed from the function?

<details>
<summary>code with prints</summary>

```

use lapin::{options::ExchangeDeclareOptions, Channel, Connection, ExchangeKind};
pub use testcontainers::{clients::Cli, core::WaitFor, Container, GenericImage, RunnableImage};

const PORT: u16 = 5672;

#[tokio::main]
async fn main() {
    // expect this to fail as no such exchange exists
    // we also expect the channel to be dead now
    let (con, channel) = setup().await;
    println!("Channel state after setup:{:?}", channel.status().state());
    channel
        .exchange_declare(
            "foo-exchange",
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                passive: true,
                ..<_>::default()
            },
            <_>::default(),
        )
        .await
        .expect("declare exchange");

}

async fn setup() -> (Connection, Channel) {
    env_logger::init();

    let docker = Cli::default();
    let container = docker.run(
        GenericImage::new("rabbitmq", "3.12-management")
            .with_wait_for(WaitFor::message_on_stdout("Server startup complete;"))
            .with_exposed_port(PORT),
    );
    println!("Connect with rabbit");
    let port = container.get_host_port_ipv4(PORT);
    let con = Connection::connect(&format!("amqp://127.0.0.1:{port}/%2f"), <_>::default())
        .await
        .unwrap();

    let channel = con.create_channel().await.expect("channel1");
    println!("Channel state during setup:{:?}", channel.status().state());

    (con, channel)
}


```

</details>

with 2.3.1 I get

```

Connect with rabbit
Channel state during setup:Connected
Channel state after setup:Error
thread 'main' panicked at 'declare exchange: InvalidChannelState(Error)', src/main.rs:24:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


```

**Comment by Keruspe** [comment]
The title references an error from an old lapin version that is no longer present in latest version.

The new logs you just posted are not the same as the ones in the initial message, and it's not the error described in the title either.

Running it again with RUST_LOG=trace for example will give more context

**Comment by alekspickle** [comment]
> The title references an error from an old lapin version that is no longer present in latest version
I am a bit confused about this, because as far as I can tell I am using the latest tag

>The new logs you just posted are not the same as the ones in the initial message, and it's not the error described in the title either.

That's what I meant when I was asking if the root cause is basically invisible in case of panic now...
essentially this is the same code


here is the trace output
<details>
<summary>logs</summary>


```

 RUST_LOG=trace cargo run
   Compiling lapin-test v0.1.0 (/home/pickle/Documents/tmh/lapin-test)
    Finished dev [unoptimized + debuginfo] target(s) in 1.52s
     Running `target/debug/lapin-test`
[2023-10-06T12:24:44Z DEBUG testcontainers::clients::cli] Executing command: "docker" "run" "--expose=5672" "-P" "-d" "rabbitmq:3.12-management"
[2023-10-06T12:24:44Z TRACE testcontainers::clients::cli] Registering starting of container c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292 at Instant { tv_sec: 23139, tv_nsec: 939660955 }
[2023-10-06T12:24:44Z DEBUG testcontainers::clients::cli] Waiting for container c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292 to be ready
[2023-10-06T12:24:44Z TRACE testcontainers::clients::cli] Time since container c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292 was started: Some(64.478µs)
[2023-10-06T12:24:47Z INFO  testcontainers::core::logs] Found message after comparing 177 lines
[2023-10-06T12:24:47Z DEBUG testcontainers::clients::cli] Container c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292 is now ready!
[2023-10-06T12:24:47Z TRACE testcontainers::clients::cli] Fetched container info: ContainerInspectResponse {
        id: Some(
            "c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292",
        ),
        created: Some(
            "2023-10-06T12:24:44.358491622Z",
        ),
        path: Some(
            "docker-entrypoint.sh",
        ),
        args: Some(
            [
                "rabbitmq-server",
            ],
        ),
        state: Some(
            ContainerState {
                status: Some(
                    RUNNING,
                ),
                running: Some(
                    true,
                ),
                paused: Some(
                    false,
                ),
                restarting: Some(
                    false,
                ),
                oom_killed: Some(
                    false,
                ),
                dead: Some(
                    false,
                ),
                pid: Some(
                    138853,
                ),
                exit_code: Some(
                    0,
                ),
                error: Some(
                    "",
                ),
                started_at: Some(
                    "2023-10-06T12:24:44.733606079Z",
                ),
                finished_at: Some(
                    "0001-01-01T00:00:00Z",
                ),
                health: None,
            },
        ),
        image: Some(
            "sha256:1f11da51bb48942a386d6de8fbfd0ca7a6537b653c56db1dced845ad43bc3b46",
        ),
        resolv_conf_path: Some(
            "/var/lib/docker/containers/c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292/resolv.conf",
        ),
        hostname_path: Some(
            "/var/lib/docker/containers/c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292/hostname",
        ),
        hosts_path: Some(
            "/var/lib/docker/containers/c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292/hosts",
        ),
        log_path: Some(
            "/var/lib/docker/containers/c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292/c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292-json.log",
        ),
        name: Some(
            "/modest_cori",
        ),
        restart_count: Some(
            0,
        ),
        driver: Some(
            "overlay2",
        ),
        platform: Some(
            "linux",
        ),
        mount_label: Some(
            "",
        ),
        process_label: Some(
            "",
        ),
        app_armor_profile: Some(
            "docker-default",
        ),
        exec_ids: None,
        host_config: Some(
            HostConfig {
                cpu_shares: Some(
                    0,
                ),
                memory: Some(
                    0,
                ),
                cgroup_parent: Some(
                    "",
                ),
                blkio_weight: Some(
                    0,
                ),
                blkio_weight_device: Some(
                    [],
                ),
                blkio_device_read_bps: Some(
                    [],
                ),
                blkio_device_write_bps: Some(
                    [],
                ),
                blkio_device_read_iops: Some(
                    [],
                ),
                blkio_device_write_iops: Some(
                    [],
                ),
                cpu_period: Some(
                    0,
                ),
                cpu_quota: Some(
                    0,
                ),
                cpu_realtime_period: Some(
                    0,
                ),
                cpu_realtime_runtime: Some(
                    0,
                ),
                cpuset_cpus: Some(
                    "",
                ),
                cpuset_mems: Some(
                    "",
                ),
                devices: Some(
                    [],
                ),
                device_cgroup_rules: None,
                device_requests: None,
                kernel_memory: None,
                kernel_memory_tcp: None,
                memory_reservation: Some(
                    0,
                ),
                memory_swap: Some(
                    0,
                ),
                memory_swappiness: None,
                nano_cpus: Some(
                    0,
                ),
                oom_kill_disable: None,
                init: None,
                pids_limit: None,
                ulimits: None,
                cpu_count: Some(
                    0,
                ),
                cpu_percent: Some(
                    0,
                ),
                io_maximum_iops: Some(
                    0,
                ),
                io_maximum_bandwidth: Some(
                    0,
                ),
                binds: None,
                container_id_file: Some(
                    "",
                ),
                log_config: Some(
                    HostConfigLogConfig {
                        typ: Some(
                            "json-file",
                        ),
                        config: Some(
                            {},
                        ),
                    },
                ),
                network_mode: Some(
                    "default",
                ),
                port_bindings: Some(
                    {},
                ),
                restart_policy: Some(
                    RestartPolicy {
                        name: Some(
                            NO,
                        ),
                        maximum_retry_count: Some(
                            0,
                        ),
                    },
                ),
                auto_remove: Some(
                    false,
                ),
                volume_driver: Some(
                    "",
                ),
                volumes_from: None,
                mounts: None,
                cap_add: None,
                cap_drop: None,
                cgroupns_mode: Some(
                    PRIVATE,
                ),
                dns: Some(
                    [],
                ),
                dns_options: Some(
                    [],
                ),
                dns_search: Some(
                    [],
                ),
                extra_hosts: None,
                group_add: None,
                ipc_mode: Some(
                    "private",
                ),
                cgroup: Some(
                    "",
                ),
                links: None,
                oom_score_adj: Some(
                    0,
                ),
                pid_mode: Some(
                    "",
                ),
                privileged: Some(
                    false,
                ),
                publish_all_ports: Some(
                    true,
                ),
                readonly_rootfs: Some(
                    false,
                ),
                security_opt: None,
                storage_opt: None,
                tmpfs: None,
                uts_mode: Some(
                    "",
                ),
                userns_mode: Some(
                    "",
                ),
                shm_size: Some(
                    67108864,
                ),
                sysctls: None,
                runtime: Some(
                    "runc",
                ),
                console_size: Some(
                    [
                        0,
                        0,
                    ],
                ),
                isolation: Some(
                    EMPTY,
                ),
                masked_paths: Some(
                    [
                        "/proc/asound",
                        "/proc/acpi",
                        "/proc/kcore",
                        "/proc/keys",
                        "/proc/latency_stats",
                        "/proc/timer_list",
                        "/proc/timer_stats",
                        "/proc/sched_debug",
                        "/proc/scsi",
                        "/sys/firmware",
                    ],
                ),
                readonly_paths: Some(
                    [
                        "/proc/bus",
                        "/proc/fs",
                        "/proc/irq",
                        "/proc/sys",
                        "/proc/sysrq-trigger",
                    ],
                ),
            },
        ),
        graph_driver: Some(
            GraphDriverData {
                name: "overlay2",
                data: {
                    "MergedDir": "/var/lib/docker/overlay2/4c23c2dde88dc8bb04e66c09e5471c497bb5a16411752fc51a947dd6bfbfe3b5/merged",
                    "WorkDir": "/var/lib/docker/overlay2/4c23c2dde88dc8bb04e66c09e5471c497bb5a16411752fc51a947dd6bfbfe3b5/work",
                    "LowerDir": "/var/lib/docker/overlay2/4c23c2dde88dc8bb04e66c09e5471c497bb5a16411752fc51a947dd6bfbfe3b5-init/diff:/var/lib/docker/overlay2/53aec778bbbcbdcedd76975d620378b5bd9b9d217be1680b1ba816b29be6b5b7/diff:/var/lib/docker/overlay2/82c86005ebcd713aefcf52e72db750e48cf8854dad85b40f51799e57a33dd824/diff:/var/lib/docker/overlay2/2cf8a95333cdc15b4c977692a4546985c2e59faf91a415aadb06a9dd746d5389/diff:/var/lib/docker/overlay2/d7c40db9e93d4854976426b68e39a7c4906af79a00921344513823e82f8b695a/diff:/var/lib/docker/overlay2/2c8e58e5f1966e734aafee6ad5818bcd8ad9924e4d51e6fdd43dc0486ee8db19/diff:/var/lib/docker/overlay2/a6605493416db76515ad81af02c6722fae2e60e965a3173e8b755604aca4f9ee/diff:/var/lib/docker/overlay2/040303b583f3419da863b040b1bfd823ad731997ae2ed9d3fe8231862ab4f36e/diff:/var/lib/docker/overlay2/89d79f5c33b63c5fa3ef74da4015b3d35024428895504c8b39c318b724a46420/diff:/var/lib/docker/overlay2/e35cd5e7e0edb6343a9bb599b1c87fd4fdf3d13af55ebdf58408003aa55ce8e5/diff:/var/lib/docker/overlay2/7115960f51ad57739cb2dccad5935e112c7820875e1c734e878cbc7272e113df/diff:/var/lib/docker/overlay2/c44aea186f40655209123b063d921645104a4d8789446e21ebb8779f1dec2b3f/diff",
                    "UpperDir": "/var/lib/docker/overlay2/4c23c2dde88dc8bb04e66c09e5471c497bb5a16411752fc51a947dd6bfbfe3b5/diff",
                },
            },
        ),
        size_rw: None,
        size_root_fs: None,
        mounts: Some(
            [
                MountPoint {
                    typ: Some(
                        VOLUME,
                    ),
                    name: Some(
                        "34746a35295f55d50f67ef04fe50db30c503ad775e4e4242220b5ddc8ac427d0",
                    ),
                    source: Some(
                        "/var/lib/docker/volumes/34746a35295f55d50f67ef04fe50db30c503ad775e4e4242220b5ddc8ac427d0/_data",
                    ),
                    destination: Some(
                        "/var/lib/rabbitmq",
                    ),
                    driver: Some(
                        "local",
                    ),
                    mode: Some(
                        "",
                    ),
                    rw: Some(
                        true,
                    ),
                    propagation: Some(
                        "",
                    ),
                },
            ],
        ),
        config: Some(
            ContainerConfig {
                hostname: Some(
                    "c696dd253ddc",
                ),
                domainname: Some(
                    "",
                ),
                user: Some(
                    "",
                ),
                attach_stdin: Some(
                    false,
                ),
                attach_stdout: Some(
                    false,
                ),
                attach_stderr: Some(
                    false,
                ),
                exposed_ports: Some(
                    {
                        "15692/tcp": {},
                        "15691/tcp": {},
                        "25672/tcp": {},
                        "5672/tcp": {},
                        "15672/tcp": {},
                        "4369/tcp": {},
                        "15671/tcp": {},
                        "5671/tcp": {},
                    },
                ),
                tty: Some(
                    false,
                ),
                open_stdin: Some(
                    false,
                ),
                stdin_once: Some(
                    false,
                ),
                env: Some(
                    [
                        "PATH=/opt/rabbitmq/sbin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                        "RABBITMQ_DATA_DIR=/var/lib/rabbitmq",
                        "RABBITMQ_VERSION=3.12.1",
                        "RABBITMQ_PGP_KEY_ID=0x0A9AF2115F4687BD29803A206B73A36E6026DFCA",
                        "RABBITMQ_HOME=/opt/rabbitmq",
                        "HOME=/var/lib/rabbitmq",
                        "LANG=C.UTF-8",
                        "LANGUAGE=C.UTF-8",
                        "LC_ALL=C.UTF-8",
                    ],
                ),
                cmd: Some(
                    [
                        "rabbitmq-server",
                    ],
                ),
                healthcheck: None,
                args_escaped: None,
                image: Some(
                    "rabbitmq:3.12-management",
                ),
                volumes: Some(
                    {
                        "/var/lib/rabbitmq": {},
                    },
                ),
                working_dir: Some(
                    "",
                ),
                entrypoint: Some(
                    [
                        "docker-entrypoint.sh",
                    ],
                ),
                network_disabled: None,
                mac_address: None,
                on_build: None,
                labels: Some(
                    {
                        "org.opencontainers.image.ref.name": "ubuntu",
                        "org.opencontainers.image.version": "22.04",
                    },
                ),
                stop_signal: None,
                stop_timeout: None,
                shell: None,
            },
        ),
        network_settings: Some(
            NetworkSettings {
                bridge: Some(
                    "",
                ),
                sandbox_id: Some(
                    "c1b49e61e7f769be1f0fa6c28e9d4fa4b2e2250fe7433fd7b180eb9dd1bcbe34",
                ),
                hairpin_mode: Some(
                    false,
                ),
                link_local_ipv6_address: Some(
                    "",
                ),
                link_local_ipv6_prefix_len: Some(
                    0,
                ),
                ports: Some(
                    {
                        "4369/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33026",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33026",
                                    ),
                                },
                            ],
                        ),
                        "15692/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33020",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33020",
                                    ),
                                },
                            ],
                        ),
                        "5671/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33025",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33025",
                                    ),
                                },
                            ],
                        ),
                        "5672/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33024",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33024",
                                    ),
                                },
                            ],
                        ),
                        "15672/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33022",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33022",
                                    ),
                                },
                            ],
                        ),
                        "25672/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33019",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33019",
                                    ),
                                },
                            ],
                        ),
                        "15691/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33021",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33021",
                                    ),
                                },
                            ],
                        ),
                        "15671/tcp": Some(
                            [
                                PortBinding {
                                    host_ip: Some(
                                        "0.0.0.0",
                                    ),
                                    host_port: Some(
                                        "33023",
                                    ),
                                },
                                PortBinding {
                                    host_ip: Some(
                                        "::",
                                    ),
                                    host_port: Some(
                                        "33023",
                                    ),
                                },
                            ],
                        ),
                    },
                ),
                sandbox_key: Some(
                    "/var/run/docker/netns/c1b49e61e7f7",
                ),
                secondary_ip_addresses: None,
                secondary_ipv6_addresses: None,
                endpoint_id: Some(
                    "db2dfc79ae1c4268c05ddc2a526c2785473080d1f1da78d86ed3f8dabbd90f47",
                ),
                gateway: Some(
                    "172.17.0.1",
                ),
                global_ipv6_address: Some(
                    "",
                ),
                global_ipv6_prefix_len: Some(
                    0,
                ),
                ip_address: Some(
                    "172.17.0.2",
                ),
                ip_prefix_len: Some(
                    16,
                ),
                ipv6_gateway: Some(
                    "",
                ),
                mac_address: Some(
                    "02:42:ac:11:00:02",
                ),
                networks: Some(
                    {
                        "bridge": EndpointSettings {
                            ipam_config: None,
                            links: None,
                            aliases: None,
                            network_id: Some(
                                "df5cb174a08fbd27a9ef190f3e359c89a55a8e16cbc740d252b15a82507f5f9f",
                            ),
                            endpoint_id: Some(
                                "db2dfc79ae1c4268c05ddc2a526c2785473080d1f1da78d86ed3f8dabbd90f47",
                            ),
                            gateway: Some(
                                "172.17.0.1",
                            ),
                            ip_address: Some(
                                "172.17.0.2",
                            ),
                            ip_prefix_len: Some(
                                16,
                            ),
                            ipv6_gateway: Some(
                                "",
                            ),
                            global_ipv6_address: Some(
                                "",
                            ),
                            global_ipv6_prefix_len: Some(
                                0,
                            ),
                            mac_address: Some(
                                "02:42:ac:11:00:02",
                            ),
                            driver_opts: None,
                        },
                    },
                ),
            },
        ),
    }
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 4369 -> 33026
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 4369 -> 33026
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 15692 -> 33020
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 15692 -> 33020
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 5671 -> 33025
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 5671 -> 33025
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 5672 -> 33024
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 5672 -> 33024
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 15672 -> 33022
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 15672 -> 33022
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 25672 -> 33019
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 25672 -> 33019
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 15691 -> 33021
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 15691 -> 33021
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv4 port mapping: 15671 -> 33023
[2023-10-06T12:24:47Z DEBUG testcontainers::core::ports] Registering IPv6 port mapping: 15671 -> 33023
Connect with rabbit
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE polling::epoll] add: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: false, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] add: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] new: epoll_fd=9, event_fd=10, timer_fd=Some(11)
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: completed
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 50 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: notified
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 50 us
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on()
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 75 us
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE polling::epoll] add: epoll_fd=9, fd=12, ev=Event { key: 0, readable: false, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 100 us
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 250 us
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=12, ev=Event { key: 0, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 500 us
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 1 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: stops hogging the reactor
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: notified
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: waiting on I/O
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=12, ev=Event { key: 0, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 1 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 50 us
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=12, ev=Event { key: 0, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE polling] Poller::notify()
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] notify: epoll_fd=9, event_fd=10
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 75 us
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: notified
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::notify()
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE polling::epoll] notify: epoll_fd=9, event_fd=10
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: notified
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 1 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=12, ev=Event { key: 0, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 100 us
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling] Poller::notify()
[2023-10-06T12:24:47Z TRACE polling::epoll] notify: epoll_fd=9, event_fd=10
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: notified
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, None)
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=None
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 250 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 500 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 750 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 1000 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 2500 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 5000 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 10000 us
[2023-10-06T12:24:47Z TRACE polling] Poller::notify()
[2023-10-06T12:24:47Z TRACE polling::epoll] notify: epoll_fd=9, event_fd=10
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: notified
[2023-10-06T12:24:47Z TRACE polling] Poller::notify()
[2023-10-06T12:24:47Z TRACE polling::epoll] notify: epoll_fd=9, event_fd=10
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, Some(29.999890742s))
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=Some(29.999890742s)
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 0 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 50 us
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, Some(29.999690794s))
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=Some(29.999690794s)
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 1 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 50 us
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=12, ev=Event { key: 0, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, Some(29.999271451s))
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=Some(29.999271451s)
Channel state during setup:Connected
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z DEBUG testcontainers::core::container] Deleting docker container c696dd253ddc7f21e61b77eb71a1a8e8d4321869ae1c0dbb65ab598d219a4292
[2023-10-06T12:24:47Z TRACE polling::epoll] new events: epoll_fd=9, res=1
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=10, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::reactor] react: 1 ready wakers
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: sleeping for 50 us
[2023-10-06T12:24:47Z TRACE polling::epoll] remove: epoll_fd=9, fd=12
[2023-10-06T12:24:47Z TRACE async_io::driver] main_loop: waiting on I/O
[2023-10-06T12:24:47Z TRACE async_io::reactor] process_timers: 0 ready wakers
[2023-10-06T12:24:47Z TRACE polling] Poller::wait(_, Some(29.967599318s))
[2023-10-06T12:24:47Z TRACE polling::epoll] wait: epoll_fd=9, timeout=Some(29.967599318s)
[2023-10-06T12:24:47Z TRACE polling::epoll] modify: epoll_fd=9, fd=11, ev=Event { key: 18446744073709551615, readable: true, writable: false }
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: notified
[2023-10-06T12:24:47Z TRACE async_io::driver] block_on: sleep until notification
Channel state after setup:Error
thread 'main' panicked at 'declare exchange: InvalidChannelState(Error)', src/main.rs:25:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

</details>

**Comment by Keruspe** [comment]
Lapin 2.0 and later are using tracing now for logging, so they don't appear here it seems with env_logger

**Comment by alekspickle** [comment]
ah, sure, lemme add that

**Comment by alekspickle** [comment]
Goddammit, the container gets scrammed when it goes out of scope, sorry for wasting your time
